Add some essential api docs

i-trofimtschuk opened the pull request:
includes history.json docs which are added in another PR (but could just be merged after that one)

please double check that it's actually the API we provide :D

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,189 @@
+                  rows:
+                    type: array
+                  groups:
+                    type: array
```
@groupsky i did not further specify the type/shapes here, but could easily be added/extended if needed

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,189 @@
+  version: 1.0.0
+
+paths:
+#  /:
+#  /{queue_name}:
+#  /{queue_name}/{page}:
+#  /history:
```
```suggestion
```

groupsky commented on docs/api.yml:
```diff
@@ -0,0 +1,185 @@
+                  pagination:
+                    type: object
+                    properties:
+                      pages_in_window:
+                        type: integer
```
What does this number represent?

groupsky replied:
current production api doesn't return number but 
```json
{
  "pagination": {
    "pages_in_window": [
      { "number": 1, "url": "/admin/rt_dashboard/inner/%5Bfailed%5D/1" }
    ]
  }
}
```

i-trofimtschuk replied:
ok i'll have to check

groupsky commented on docs/api.yml:
```diff
@@ -0,0 +1,185 @@
+                    properties:
+                      pages_in_window:
+                        type: integer
+                      next_page:
+                        type: integer
```
I guess this is the index of next page - could it be different than `cur_page + 1`? In either case it's good to have a quick info about it

i-trofimtschuk replied:
can check / add, yes

i-trofimtschuk commented:
@groupsky i think i fixed both issues in the api docs, how did you verify it easily? can you recheck?

vasilty commented on docs/api.yml:
```diff
@@ -0,0 +1,215 @@
+              schema:
+                type: object
+                properties:
+                  rows:
```
`groups` property is missing

groupsky replied:
Should we specify it if it's going to be removed?

i-trofimtschuk replied:
added, reflecting https://github.com/djangsters/rt-dashboard/pull/40/files?diff=split&w=1#diff-16147ada3f9dd87580394636cdb94468R116-R117

see below

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,215 @@
+                          type: string
+                          enum: [range, point]
+                        content:
+                          type: string
```
```suggestion
                          type: string
                  groups:
                    type: array
                    items:
                      type: object
                      properties:
                        id:
                          type: string
                        content:
                          type: number
                        order:
                          type: number
```

vasilty commented on docs/api.yml:
```diff
@@ -0,0 +1,226 @@
+                          type: number
+                        start:
+                          type: number
+                        end:
```
`end` might not be present https://github.com/djangsters/rt-dashboard/blob/7008ffdb55db848f5882957de16725ba5748c279/rt_dashboard/history.py#L100

groupsky replied:
`end` is not a required field

vasilty commented on docs/api.yml:
```diff
@@ -0,0 +1,226 @@
+                    items:
+                      type: object
+                      properties:
+                        group:
```
`title` property is missing https://github.com/djangsters/rt-dashboard/blob/7008ffdb55db848f5882957de16725ba5748c279/rt_dashboard/history.py#L84

vasilty replied:
`style` property is missing https://github.com/djangsters/rt-dashboard/blob/7008ffdb55db848f5882957de16725ba5748c279/rt_dashboard/history.py#L106

groupsky replied:
updated both and added decriptions

vasilty commented on docs/api.yml:
```diff
@@ -0,0 +1,226 @@
+                    type: string
+  /queue/{queue_name}/empty:
+    post:
+      summary: Empty the queue with the specified queue_name
```
`[failed]` and `[finished]` and not queues but registries. Do you think we should mention that?

i-trofimtschuk replied:
i think this belongs in the usage docs, i also mentioned that failed/finished/runing are registries in the queues.json endpoint, i think this is enough for api docs

vasilty commented on docs/api.yml:
```diff
@@ -0,0 +1,226 @@
+      responses:
+        200:
+          description: |
+            Return object containting list of queues and theis states
```
I'm not sure description gives more insight here than the summary

i-trofimtschuk replied:
```suggestion
            Return object containing a list of queues and the `[failed]`,
            `[finished]` and `[running]` registries, the number of elements
            in them and an url for more details on them.
```

vasilty commented on docs/api.yml:
```diff
@@ -0,0 +1,226 @@
+                          format: uuid
+                        status:
+                          type: string
+                        enqueued_at:
```
`ended_at` is missing https://github.com/djangsters/rt-dashboard/blob/7008ffdb55db848f5882957de16725ba5748c279/rt_dashboard/web.py#L123

vasilty replied:
for running jobs we also return `worker` https://github.com/djangsters/rt-dashboard/blob/7008ffdb55db848f5882957de16725ba5748c279/rt_dashboard/web.py#L281

vasilty commented on docs/api.yml:
```diff
@@ -0,0 +1,226 @@
+                  name:
+                    type: string
+                  jobs:
+                    type: array
```
I think we should mention sorting of returned jobs here

i-trofimtschuk replied:
```suggestion
                    type: array
                    description: |
                      Default ordering is oldest entries first, except for the
                      `[failed]` and `[finished]` registries, where it is newest first.
                    
```

vasilty commented on docs/api.yml:
```diff
@@ -0,0 +1,226 @@
+                                url:
+                                    type: string
+                                    format: url
+                      next_page:
```
We might want to change/comment that `next_page` is actually the last page https://github.com/djangsters/rt-dashboard/blob/7008ffdb55db848f5882957de16725ba5748c279/rt_dashboard/web.py#L259

vasilty commented on docs/api.yml:
```diff
@@ -0,0 +1,226 @@
+                            url:
+                                type: string
+                                format: url
+                      prev_page:
```
And `prev_page` is actually the first page https://github.com/djangsters/rt-dashboard/blob/7008ffdb55db848f5882957de16725ba5748c279/rt_dashboard/web.py#L255

vasilty commented:
Can we somehow generate docs automatically or validate the documentation? Even for relatively easy APIs it is difficult to create and manage docs manually

groupsky commented on docs/api.yml:
```diff
@@ -0,0 +1,226 @@
+                          type: number
+                        order:
+                          type: number
+  /job/{job_id}/cancel:
+    post:
+      summary: Cancel the job with the specified job_id
+      parameters:
+        - name: job_id
+          in: path
+          schema:
+            type: string
+            format: uuid
+          required: true
```
does not mention the csrf requirement

i-trofimtschuk replied:
@groupsky CSRF protection is a django specific thing ^^

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -20,20 +20,38 @@ paths:
+                        - group
+                        - subgroup
+                        - start
+                        - title
```
```suggestion
                        - title
                        - type
                        - content
```

i-trofimtschuk replied:
@groupsky type and content are also always present

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,244 @@
+                          type: string
+                          description: Short info about the task
+                        title:
+                          type: string
```
```suggestion
                          type: string
                          format: html
```

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,253 @@
+                        enqueued_at:
+                          type: string
+                          format: date-time
+                        started_at:
+                          type: string
+                          format: date-time
```
```suggestion
                        started_at:
                          type: string
                          format: date-time
                        ended_at:
                          type: string
                          format: date-time
```

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,256 @@
+                      `[failed]` and `[finished]` registries, where it is newest first.
+                    
+                    items:
+                      type: object
```
```suggestion
                      type: object
                      required: 
                        - id
                        - status
                        - enqueued_at
                        - started_at
                        - ended_at
                        - origin
                        - error_message
                        - description
```

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,265 @@
+                        error_message:
+                          type: string
+                        description:
+                          type: string
```
```suggestion
                          type: string
                        worker:
                          type: string
                          description: |
                            Description of Worker that the job is being executed on.
                            Only present for running jobs (aka in `[running]` registry)
```

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,270 @@
+                            url:
+                              type: string
+                              format: url
+                      next_page:
+                        type: object
+                        properties:
+                          url:
+                            type: string
+                            format: url
+                      prev_page:
+                        type: object
+                        properties:
+                          url:
+                            type: string
+                            format: url
```
```suggestion
                      next_page:
                        type: object
                        nullable: true
                        description: |
                          Despite the naming actually provides a link to the LAST page.
                          On the last page 'null' is used instead of an object.
                        properties:
                          url:
                            type: string
                            format: url
                      prev_page:
                        type: object
                        nullable: true
                        description: |
                          Despite the naming actually provides a link to the FIRST page.
                          On the first page 'null' is used instead of an object.
                        properties:
                          url:
                            type: string
                            format: url
```

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,278 @@
+          content:
+            application/json:
+              schema:
+                type: object
```
```suggestion
                type: object
                required:
                  - queues
```

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,278 @@
+                  queues:
+                    type: array
+                    items:
+                      type: object
```
```suggestion
                      type: object
                      required:
                        - name
                        - count
                        - url
```

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,284 @@
+                    description: |
+                      Default ordering is oldest entries first, except for the
+                      `[failed]` and `[finished]` registries, where it is newest first.
+                    
```
```suggestion
```

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,283 @@
+          content:
+            application/json:
+              schema:
+                type: object
```
```suggestion
                type: object
                required:
                  - name
                  - jobs
                  - pagination
```

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,283 @@
+                            Description of Worker that the job is being executed on.
+                            Only present for running jobs (aka in `[running]` registry)
+                  pagination:
+                    type: object
```
```suggestion
                    type: object
                    required:
                      - pages_in_window
                      - next_page
                      - prev_page
```

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,291 @@
+          content:
+            application/json:
+              schema:
+                type: object
```
```suggestion
                type: object
                required:
                  - workers
```

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,291 @@
+                  workers:
+                    type: array
+                    items:
+                      type: object
```
```suggestion
                      type: object
                      required:
                        - name
                        - queues
                        - state
```

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,297 @@
+                        - state
+                      properties:
+                        name:
+                          type: string
```
```suggestion
                          type: string
                          description: Description of the worker
```

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,297 @@
+                        name:
+                          type: string
+                        queues:
+                          type: array
```
```suggestion
                          type: array
                          description: List of queue names the worker consumes from
```

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,297 @@
+                          items:
+                            type: string
+                        state:
+                          type: string
```
```suggestion
                          type: string
                          enum: [idle, busy, dead]
```

i-trofimtschuk commented:
@vasilty @groupsky i think i addressed all comments now and tried to incorporate/extend it to be in line with the code, care to double check?
i'd like to have this merged soonish even if it's not 100% complete, we can still extend it.
what do you think?

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,300 @@
+          content:
+            application/json:
+              schema:
+                type: object
```
```suggestion
                type: object
                required:
                  - rows
                  - groups
```

i-trofimtschuk commented on docs/api.yml:
```diff
@@ -0,0 +1,300 @@
+                  groups:
+                    type: array
+                    items:
+                      type: object
```
```suggestion
                      type: object
                      required:
                        - id
                        - content
                        - order
```

vasilty commented on docs/api.yml:
```diff
@@ -0,0 +1,307 @@
+                          enum: [range, point]
+                        content:
+                          type: string
+                          description: Short info about the task
```
Info about the worker the task is running on

i-trofimtschuk replied:
```suggestion
                          description: |
                            Info about the worker the task is running on,
                            or the time range string, or the starting time string,
                            depending on the state
```

vasilty commented on docs/api.yml:
```diff
@@ -0,0 +1,307 @@
+          description: |
+            Return object containing a list of queues and the `[failed]`,
+            `[finished]` and `[running]` registries, the number of elements
+            in them and an url for more details on them.
```
I googled, found a not very reputable source but I agree
>Although the U at the first of URL is a vowel, the abbreviation is pronounced “you-are-ell.” Because the letter Y is a consonant in this case, use a rather than an.

i-trofimtschuk replied:
```suggestion
            in them and a URL for more details on them.
```

vasilty commented on docs/api.yml:
```diff
@@ -0,0 +1,307 @@
+                        nullable: true
+                        description: |
+                          Despite the naming actually provides a link to the LAST page.
+                          On the last page 'null' is used instead of an object.
```
```suggestion
                          On the last page 'null' is returned instead of an object.
```

vasilty commented on docs/api.yml:
```diff
@@ -0,0 +1,307 @@
+                        nullable: true
+                        description: |
+                          Despite the naming actually provides a link to the FIRST page.
+                          On the first page 'null' is used instead of an object.
```
```suggestion
                          On the first page 'null' is returned instead of an object.
```

dkusnik commented on docs/api.yml:
```diff
@@ -0,0 +1,307 @@
+                  status:
+                    type: string
+  /queue/{queue_name}/delete:
+    post:
```
why post instead of delete?

i-trofimtschuk replied:
@dkusnik just documenting the state here ^^
